### PR TITLE
TST: Check requires_memory immediately before the test

### PR DESCRIPTION
**Pull request by pv**
Perform the available-memory check immediately before the test, not at
import time, and catch MemoryError.

Possibly helps with #15071

### CODE REVIEW

File: numpy/testing/_private/utils.py [hunk 1]
```diff
@@ -2388,6 +2388,22 @@ def requires_memory(free_bytes):
...
     import pytest
 
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*a, **kw):
+            msg = check_free_memory(free_bytes)
+            if msg is not None:
+                pytest.skip(msg)
+            return func(*a, **kw)
+        return wrapper
```
Comment (charris) [inline numpy/testing/_private/utils.py]: Be good to put a blank line after the function decorator.

File: numpy/testing/_private/utils.py [hunk 2]
```diff
@@ -2388,6 +2388,22 @@ def requires_memory(free_bytes):
     """Decorator to skip a test if not enough memory is available"""
     import pytest
 
+    def decorator(func):
+        @wraps(func)
+        def wrapper(*a, **kw):
+            msg = check_free_memory(free_bytes)
+            if msg is not None:
+                pytest.skip(msg)
+            return func(*a, **kw)
```
Comment (charris) [inline numpy/testing/_private/utils.py]: Might be worth catching `MemoryError` here as an extra check.

Comment (pv) [reply numpy/testing/_private/utils.py]: Done.

File: numpy/testing/_private/utils.py (hunk 1 above)
Comment (pv) [reply numpy/testing/_private/utils.py]: ok

**Comment by charris** [comment]
Thanks Pauli.

**Comment by charris** [comment]
That fixed the wheels build, the test now xfails. I wonder why the memory check is not succeeding?

**Comment by pv** [comment]
The xfail count on travis in this repo is 16, so I think it doesn't fail there. The tests in numpy-wheels iiuc are run inside docker, and maybe that distorts numbers (cf. [here](https://gitlab.com/procps-ng/procps/blob/24fd2605c51fccc375ab0287cec33aa767f06718/proc/sysinfo.c#L765)). Seems to be a bit cursed...

**Comment by charris** [comment]
> Seems to be a bit cursed...

Indeed. Good to know.

**Comment by charris** [comment]
Seems we could dispense with the environmental variable and just rely on the xfail.

**Comment by pv** [comment]
Alternative (1): drop the custom meminfo code, and rely on psutil getting it right. Psutil is a test dependency and should be installed on CIs.

Alternative (2): as above, but keep the meminfo code and fix it. Psutil should still be installed on CIs.

Alternative (3): give up trying to measure memory, just rely on env var. If not set, skip tests.

Alternative (4): try and catch MemoryError. In some cases with swap, the machine may hang for some time (as occurs for me with the blas64 test).

Alternative (5): as (4) but also have the env var for skipping tests.

Alternative (X): revert, and we'll think about this later...

**Pull request merged**